lib/iris/tests/unit/coords/test_DimCoord.py

```diff
@@ -0,0 +1,620 @@
+# (C) British Crown Copyright 2013 - 2017, Met Office
+#
+# This file is part of Iris.
+#
+# Iris is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Iris is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Iris.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Unit tests for the :class:`iris.coords.DimCoord` class.
+
+Note: a lot of these methods are actually defined by the :class:`Coord` class,
+but can only be tested on concrete DimCoord/DimCoord instances.
+In addition, the DimCoord class has little behaviour for some of these, as it
+cannot contain lazy points or bounds data.
+
+"""
+
+from __future__ import (absolute_import, division, print_function)
+from six.moves import (filter, input, map, range, zip)  # noqa
+
+# Import iris.tests first so that some things can be initialised before
+# importing anything else.
+import iris.tests as tests
+
+import dask.array as da
+import mock
+import numpy as np
+import unittest
+
+from iris.coords import DimCoord
+
+
+_DEBUG = True
+
+
+_DIMCOORD_LAZY_API = hasattr(DimCoord, 'has_lazy_points')
+_DEBUG_NONLAZY_CONTROL = _DEBUG
+if _DEBUG_NONLAZY_CONTROL:
+    print()
+    print('LAZY COORD API DETECT = ', _DIMCOORD_LAZY_API)
+    print()
+
+
+def skip_lazy_auxcoords(fn):
+    """
+    Decorator to choose whether to run tests, based on the API version of
+    the DimCoord class.
+
+    Example usage:
+        @skip_lazy_coord
+        class TestMyStuff(tests.IrisTest):
+            ...
+
+    """
+    skip = unittest.skipIf(
+        condition=not _DIMCOORD_LAZY_API,
+        reason=('Test(s) only applicable to the enhanced DimCoord API '
+                '(lazy support).'))
+
+    return skip(fn)
+
+
+# Utility functions to get underlying points/bounds data, which work with old
+# coords as well as new.
+def _core_points(coord):
```

Comment:
I will consider moving common code into `iris.tests.unit.coords`, but it will undoubtedly make them less readable.